core/store/src/trie/trie_storage.rs

```diff
@@ -184,31 +220,53 @@ impl TrieCachingStorage {
     fn inc_counter(&self) {
         self.counter.set(self.counter.get() + 1);
     }
+
+    /// If we are in the caching chunk state, put item to the chunk cache.
+    fn try_put_to_chunk_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        if let TrieCacheState::CachingChunk = self.cache_state.borrow().get() {
+            self.chunk_cache.borrow_mut().insert(key, value);
+        };
+    }
+
+    /// Set cache state.
+    pub fn set_state(&self, state: TrieCacheState) {
+        self.cache_state.set(state);
+    }
 }
 
 impl TrieStorage for TrieCachingStorage {
     fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
+        // Try to get value from chunk cache containing free of charge nodes, otherwise increment the nodes counter.
+        // let RawBytesWithCost { value, cost } = self.get_from_cache(hash);
+        if let Some(val) = self.chunk_cache.borrow_mut().get(hash) {
+            return Ok(val.clone());
+        }
         self.inc_counter();
-        let mut guard = self.cache.0.lock().expect(POISONED_LOCK_ERR);
-        if let Some(val) = guard.get(hash) {
-            Ok(val.clone())
-        } else {
-            let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
-            let val = self
-                .store
-                .get(ColState, key.as_ref())
-                .map_err(|_| StorageError::StorageInternalError)?;
-            if let Some(val) = val {
-                let val: Arc<[u8]> = val.into();
-                if val.len() < TRIE_LIMIT_CACHED_VALUE_SIZE {
-                    guard.put(*hash, val.clone());
-                }
-                Ok(val)
-            } else {
-                // not StorageError::TrieNodeMissing because it's only for TrieMemoryPartialStorage
-                Err(StorageError::StorageInconsistentState("Trie node missing".to_string()))
-            }
+
+        // If full cost for retrieving the node must be charged, increment the counter.
+        if let Some(val) = self.shard_cache.get(hash) {
```

Comment:
I think you should acquire lock in this function instead of inside the `get` and `put` functions because you want the actions in this function to be atomic. In fact, this struct is no longer thread safe with this modification because multiple threads can modify `self.chunk_cache` at the same time, and the `get` and `put` method should be guarded by the same lock, otherwise another thread could change the value between the `get` and `put` call.

It's possible that this function is still safe because how we share cache among different threads. I just feel that the story of lock here is very confusing and need to be flushed out more.

Reply:
It is safe, because each key is a value hash, so there can be only one possible value for the given hash.

On the other hand, this change was not intentional and was caused by sequence of refactoring. I agree that it is confusing here, so I returned acquiring lock here again.